compiler/src/dotty/tools/dotc/transform/localopt/FormatChecker.scala

```diff
@@ -0,0 +1,247 @@
+package dotty.tools.dotc
+package transform.localopt
+
+import scala.annotation.tailrec
+import scala.collection.mutable.{ListBuffer, Stack}
+import scala.reflect.{ClassTag, classTag}
+import scala.util.chaining.*
+import scala.util.matching.Regex.Match
+
+import java.util.{Calendar, Date, Formattable}
+
+import StringContextChecker.InterpolationReporter
+
+/** Formatter string checker. */
+abstract class FormatChecker(using reporter: InterpolationReporter):
+
+  // Pick the first runtime type which the i'th arg can satisfy.
+  // If conversion is required, implementation must emit it.
+  def argType(argi: Int, types: ClassTag[?]*): ClassTag[?]
```

Comment:
I noticed that `FormatChecker` is introduced as an abstract class and as I understood it was done to have an ability to have a unit test with specific implementation. It also required to use `ClassTag` instead of `Type` as it would require proper context initialization. Did I get it right?

What do you think about moving all logic into a single class?
That would allow to use `Type` in `Conversion`s and other places and the part with mapping `ClassTag` to `Type` won't be required. All logic will be located in a single file that is an additional bonus.

I understand that it will be hard to keep `FormatCheckerTest` but it seems that at this moment when we got parity with migrated tests from Scala2 compiler this unit test won't be so useful as at development moment.

Reply:
Actually, I want to backport to Scala 2, so a goal was for some target independence. That's useful for any future fixes or improvements.

`ClassTag` sufficed. I think I started by abstracting over `Type` but that was baroque.

I'll take another look at the class structure. I remember an open issue was optimization or at least constant folding; not sure if that is better solved here or generally. Other optimizations would be along the lines of "replace %d with .toInt" or whatever.